Plugins: Add support for templates

luispadron opened the pull request:
Resolves https://github.com/tuist/tuist/issues/2648
Request for comments document (if applies):

### Short description 📝

This PR adds support for defining templates in plugins. Plugins will now allow adding templates to `Templates/` directory relative to the `Plugin.swift`. From there templates will be loaded and can be used with commands like `tuist scaffold`.

### Checklist ✅

- [X] The code architecture and patterns are consistent with the rest of the codebase.
- [X] The changes have been tested following the [documented guidelines](https://tuist.io/docs/contribution/testing-strategy/).
- [X] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [X] In case the PR introduces changes that affect users, the documentation has been updated.

pepibumur commented on Sources/TuistKit/Services/ScaffoldService.swift:
```diff
@@ -136,6 +144,17 @@ class ScaffoldService {
+    /// - Parameter path: The path where the command is being executed.
+    /// - Returns: A list of template directories
+    private func locateTemplateDirectories(at path: AbsolutePath) throws -> [AbsolutePath] {
+        let config = try configLoader.loadConfig(path: path)
```
Nitpick: What about moving the loading of the manifests to the `run` method and then pass it down? If we do it in private methods, we risk doing the same in another method and therefore doing unnecessary loads.

pepibumur reviewed (approved):
This is awesome @luispadron. I wouldn't have thought the PR would be so small.

fortmarek commented on Sources/TuistKit/Services/ScaffoldService.swift:
```diff
@@ -20,7 +21,7 @@ enum ScaffoldServiceError: FatalError, Equatable {
         switch self {
         case let .templateNotFound(template):
-            return "Could not find template \(template). Make sure it exists at Tuist/Templates/\(template)"
+            return "Could not find template \(template)."
```
I'd still provide information about where the template is expected. Rather than deleting the part about local template, I'd extend it with that we were not able to find it in the plugins as well.

fortmarek commented on Sources/TuistLoader/Models+ManifestMappers/Plugin+ManifestMapper.swift:
```diff
@@ -6,7 +6,10 @@ import TuistGraph
-    static func from(manifest: ProjectDescription.Plugin) throws -> Self {
+    static func from(
+        manifest: ProjectDescription.Plugin,
+        path _: AbsolutePath
```
the path is not needed as it seems ... ?

luispadron replied:
Looks the file is no longer needed 👍🏼

fortmarek commented on Sources/TuistPlugin/PluginService.swift:
```diff
@@ -16,21 +17,24 @@ public protocol PluginServicing {
     ///   - gitHandler: A git handler for cloning and interacting with remote plugins.
     public init(
         manifestLoader: ManifestLoading = ManifestLoader(),
+        templateDirectoryLocator: TemplatesDirectoryLocating = TemplatesDirectoryLocator(),
```
Nit: I'd rename `templateDirectoryLocator` to `templatesDirectoryLocator` to be consistent with the type name

fortmarek reviewed (commented):
This looks great @luispadron 🖤 ! It's awesome to see that this has not taken a lot of code to do at all.

Apart from my minor comments, I also have noticed two things when trying it out locally:
<img width="917" alt="image" src="https://user-images.githubusercontent.com/9371695/111908731-910a6780-8a5a-11eb-99fc-89ac49a6e8ed.png">

It seems that we fetch the plugins twice - can it be because we do the work once for helpers and once for templates?

It'd also be nice if local plugin templates would show in a project generated by `tuist edit` which does not seem to be the case right. Hopefully, this should not be too much additional work and would make it great impact for users developing their own plugin templates.

luispadron commented:
Good points @fortmarek about the edit, it should be simple to add i'll update the PR to include that. I'll also fix up the output

luispadron commented:
@pepibumur @fortmarek 

So it looks like when running `tuist scaffold` we first pre-process the templates in order to gather info like required and optional parems:

```swift
// TuistCommand.swift:50-52
if processedArguments.first == ScaffoldCommand.configuration.commandName {
    try ScaffoldCommand.preprocess(processedArguments)
}
```

Hence why @fortmarek receives duplicate "fetching plugins" message. I've updated this with the preprocessing in mind. The way it works is that when `ScaffoldCommand.preprocess` is called we fetch the plugins and "cache" them by maintaining:

```swift
extension ScaffoldCommand {
    static var requiredTemplateOptions: [(name: String, option: Option<String>)] = []
    static var optionalTemplateOptions: [(name: String, option: Option<String?>)] = []
    // addition
    static var plugins: Plugins = .none
}
```

This ensures we only fetch plugins once when using the scaffold command, this also ensures we know about the required template options and optional template options for plugin templates.

![image](https://user-images.githubusercontent.com/13840545/112030512-b9549d80-8b10-11eb-9737-e71b2525c109.png)


@fortmarek To the error handling point, I've updated it to output a list of "search paths" for plugins. This way as you suggested, we inform the user of where we tried to look for a template but did not find one. Let me know what you think:

![image](https://user-images.githubusercontent.com/13840545/112030646-dab58980-8b10-11eb-91c5-1f0727b0fbef.png)

fortmarek commented on Sources/TuistGraph/Models/Plugins.swift:
```diff
@@ -1,20 +1,30 @@
     ///
     /// - Parameters:
     ///     - projectDescriptionHelpers: List of the loaded helper plugins.
+    ///     - templates: List of templates loaded from plugins.
```
nit: this does not correspond to the current parameter name

I'd also rename it to `templateDirectories` to be the same as the property (or the other way around, I don't really any preferences regarding that)

fortmarek commented on Sources/TuistKit/Commands/ScaffoldCommand.swift:
```diff
@@ -94,6 +98,7 @@ struct ScaffoldCommand: ParsableCommand {
 extension ScaffoldCommand {
     static var requiredTemplateOptions: [(name: String, option: Option<String>)] = []
     static var optionalTemplateOptions: [(name: String, option: Option<String?>)] = []
+    static var plugins: Plugins = .none
```
yep, forgot about the preprocessing, good catch 👍

fortmarek commented on Sources/TuistKit/Commands/ScaffoldCommand.swift:
```diff
@@ -126,6 +134,15 @@ extension ScaffoldCommand {
+        let configLoader = ConfigLoader(manifestLoader: ManifestLoader())
+        let pluginService = PluginService()
+        let config = try configLoader.loadConfig(path: path)
+        return try pluginService.loadPlugins(using: config)
```
To make this testable, it might be better to move this logic into `ScaffoldService` as we should try to keep `ScaffoldCommand` to be as business-logic-free as possible (similarly to how we extract the options from `ScaffolService.loadTemplateOptions` function)

fortmarek commented on Sources/TuistKit/Services/ScaffoldService.swift:
```diff
@@ -13,14 +14,17 @@ enum ScaffoldServiceError: FatalError, Equatable {
         switch self {
-        case let .templateNotFound(template):
-            return "Could not find template \(template). Make sure it exists at Tuist/Templates/\(template)"
+        case let .templateNotFound(template, searchPaths):
```
current output can get pretty long if you have a lot of different templates - which could doubly be true for a plugin.

rather than having a list of `path/Templates/template_one`, `path/Templates/template_two`, etc., I think it'd be more valuable to have only the common parent directory. 

That is if I have a plugin and local templates, I'd like to see that tuist searched for them in:
`path-to-local-templates/Templates` and `path-to-plugin/Templates`.

We could probably do this by something like: 
```swift
let searchDirectories: Set<AbsolutePath> = searchPaths.reduce(into: []) { acc, path in  
   acc.insert(path.parentDirectory)
}
```

I believe this might be more helpful to the users (but correct me if you don't feel this way!)

fortmarek reviewed (commented):
Tested locally and no duplicated output 👏 I think that once the template will be in the project generated by `tuist edit`, we are good to merge.

One minor thing (a bit unrelated to this PR):
I have noticed that when for every operation that loads the plugins (`edit`, `generate`, etc.), users will always see `Downloading plugin git url: ....`. I wonder if we should maybe show this only when we actually *do* download the plugin (most of the times it should be cached locally, right?).

Users might be surprised that a network operation has to be done every time (even though it's actually not)

luispadron commented:
@fortmarek I like your suggestion about the directories:

<img width="857" alt="Screen Shot 2021-03-22 at 7 52 13 PM" src="https://user-images.githubusercontent.com/13840545/112072384-19b30180-8b48-11eb-8b8f-e35467a3609a.png">

luispadron commented:
@fortmarek Totally forgot about adding the edit support, this is now in here 👍🏼 

I also went ahead and moved the plugin loading to the `ScaffoldService` as you suggested, this means we `loadPlugins` for both pre-processing and when running the actual scaffold process. But since plugins are either cached or locally available I don't see this being an issue.

Which brings to me to the final change which is that I downgraded the logging of using local plugins and cached plugins to debug (`--verbose`) and moved the logging of when we actually fetch the plugins to when we run the `git clone` command (which is only when there is no cache). This ensures we don't have the duplicate output you experienced while still informing users that we are performing a clone when initially fetching. This will reduce the noise users see (since as you mentioned we were displaying `Fetching plugins(s)...` everytime, for every command that used plugins.

fortmarek reviewed (approved):
this is awesome 👏 this looks good to me, you might just want to run `./fourier format swift --fix` as there seems to be some issues with the linter check. Once CI is green, feel free to merge, this will be a great addition to this week's release ❤️
